src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,133 @@
+        } else {
+            throw new IllegalArgumentException("Capacity must be greater than zero.");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= array.length) {
+            grow();
+        }
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Non existent "
+                    + "position: " + index);
+        }
+        if (size >= array.length) {
+            grow();
+        }
+        Object[] result = new Object[++size];
+        System.arraycopy(array, 0, result, 0, index);
+        result[index] = value;
+        System.arraycopy(array, index, result, index + 1, size - index - 1);
+        array = result;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isValidIndex(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Not found "
+                    + "element at position: ");
+        } else {
+            return (T) array[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isValidIndex(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Not found "
+                    + "element at position: " + index);
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isValidIndex(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Not found "
+                    + "element at position: " + index);
+        }
+        Object toDelete = array[index];
+        removeAndTrim(index);
+        return (T) toDelete;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = -1;
+        for (int i = 0; i < array.length; i++) {
+            if (array[i] == null ? array[i] == element
+                    : array[i].equals(element)) {
+                index = i;
+                break;
+            }
+        }
+        if (index == -1) {
+            throw new NoSuchElementException("Not found "
+                    + "element at position: " + index);
+        }
+        Object toDelete = array[index];
+        removeAndTrim(index);
+        return (T) toDelete;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void grow() {
+        Object[] newArray = new Object[array.length + array.length / 2];
+        System.arraycopy(array, 0, newArray, 0, size);
+        array = newArray;
+    }
+
+    boolean isValidIndex(int index) {
+        return index < 0 || index >= size;
```

Comment:
The method `isValidIndex` should return `true` for valid indices. The condition should be `return index >= 0 && index < size;`.